src/main/java/tools/jackson/databind/ObjectReader.java

```diff
@@ -1320,6 +1385,207 @@ public <T> T readValue(TokenBuffer src) throws JacksonException
+     * {@link DeferredBindingException} with all problems.
+     *
+     * <p><b>Usage</b>: This method should be called on an ObjectReader created via
+     * {@link #problemCollectingReader()} or {@link #problemCollectingReader(int)}. If called on a regular
+     * reader (without problem collection enabled), it behaves the same as
+     * {@link #readValue(JsonParser)} since no handler is registered.
+     *
+     * <p><b>Error handling</b>:
+     * <ul>
+     * <li>Recoverable errors are accumulated and thrown as
+     *     {@link DeferredBindingException} after parsing</li>
+     * <li>Hard (non-recoverable) failures throw immediately, with collected problems
+     *     attached as suppressed exceptions</li>
+     * <li>When the configured limit is reached, collection stops</li>
+     * </ul>
+     *
+     * <p><b>Exception Handling Strategy</b>:
+     *
+     * <p>This method catches only {@link DatabindException} subtypes (not all
+     * {@link JacksonException}s) because:
+     *
+     * <ul>
+     * <li>Core streaming errors ({@link tools.jackson.core.exc.StreamReadException},
+     *     {@link tools.jackson.core.exc.StreamWriteException}) represent structural
+     *     JSON problems that cannot be recovered from (malformed JSON, I/O errors)</li>
+     *
+     * <li>Only databind-level errors (type conversion, unknown properties, instantiation
+     *     failures) are potentially recoverable and suitable for collection</li>
+     *
+     * <li>Catching all JacksonExceptions would hide critical parsing errors that should
+     *     fail fast</li>
+     * </ul>
+     *
+     * <p>If a hard failure occurs after some problems have been collected, those problems
+     * are attached as suppressed exceptions to the thrown exception for debugging purposes.
+     *
+     * <p><b>Thread-safety</b>: Each call allocates a fresh problem bucket,
+     * so multiple concurrent calls on the same reader instance are safe.
+     *
+     * <p><b>Parser filtering</b>: Unlike convenience overloads ({@link #readValueCollectingProblems(String)},
+     * {@link #readValueCollectingProblems(byte[])}, etc.), this method does <i>not</i> apply
+     * parser filtering. Callers are responsible for filter wrapping if needed.
+     *
+     * @param <T> Type to deserialize
+     * @param p JsonParser to read from (will not be closed by this method)
+     * @return Deserialized object
+     * @throws DeferredBindingException if recoverable problems were collected
+     * @throws DatabindException if a non-recoverable error occurred
+     * @since 3.1
+     */
+    public <T> T readValueCollectingProblems(JsonParser p) throws JacksonException {
+        _assertNotNull("p", p);
+
+        // CRITICAL: Allocate a FRESH bucket for THIS call (thread-safety)
+        List<CollectedProblem> bucket = new ArrayList<>();
+
+        // Create per-call attributes with the fresh bucket
+        ContextAttributes perCallAttrs = _config.getAttributes()
```

Comment:
Instead of this, let's make use of `DeserializationContext.setAttribute()` -- `DeserializationContext` is per-call, mutable. 

And to have access to the DC, need to refactor to have something like

    protected <T> T readValueCollectingProblems(JsonParser p, DeserializationContextExt ctxt) 

that all public `readValueCollectingProblems()` call with parser, context created.

Finally, we should NOT call `readValue(parser)` because that would just create yet another `DeserializationContext`. This means it'll be necessary to call one `_bind()` or `_bindAndClose()`.